lib/immutable-struct.rb

```diff
@@ -36,26 +37,46 @@ class ImmutableStruct
   #     p.minor    # => "yup"
   #     p.minor?   # => true
   #
+  # === Derived attributes
+  # A derived attribute is a zero-arity method (ie. getter) that applies a transformation to the
+  # existing values stored inside an +ImmutableStruct+. For instance:
+  #
+  #    ImmutableStruct.new(:foo, :bar) do
+  #      def derived; self.foo + ":" + self.bar; end
+  #    end
+  #
+  # Please note that, unless overriden, +ImmutableStruct#to_h+ will serialize all zero-arity methods.
+  #
+  #
   def self.new(*attributes,&block)
+    raise ArgumentError if attributes.empty?
     klass = Class.new do
+      array_attr = lambda { |attr| attr.kind_of?(Array) && attr.size == 1 }
+      boolean_attr = lambda { |attr| String(attr).match(/(^.*)\?$/) }
+      add_reader = lambda { |attr| begin; attr_reader attr; rescue; raise ArgumentError; end }
+
       attributes.each do |attribute|
-        if attribute.to_s =~ /(^.*)\?$/
+        case attribute
+        when String
```

Comment:
Strings work fine—any reason to disallow them?  This is technically a breaking change since strings work now.